README


This archive contains the distribution Tickit-Widget-Progressbar,
version 0.100:

  horizontal/vertical progress bars for Tickit

This software is copyright (c) 2011 by Tom Molesworth.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Tickit-Widget-Progressbar

0.100     2013-09-08 19:21:20 Europe/London
	Features:
	* Switched to Tickit::RenderBuffer
	* Helper method to expose only the areas that need drawing
	(currently disabled pending tests to prove that it's doing
	the right thing)
	* Style support
	* Initial support for gradients, currently disabled since
	it would require the Tickit::Colour mapping code which is
	not yet on CPAN.

0.002     2011-12-04 22:30:31 Europe/London
	Fix abstract, update documentation

0.001     2011-11-27 05:43:39 Europe/London
	Initial CPAN release


Makefile.PL

use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "horizontal/vertical progress bars for Tickit",
  "AUTHOR" => "Tom Molesworth <cpan\@entitymodel.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Tickit-Widget-Progressbar",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Tickit::Widget::Progressbar",
  "PREREQ_PM" => {
    "Tickit" => "0.33",
    "parent" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "Test::CheckDeps" => "0.006",
    "Test::More" => "0.98",
    "blib" => 0
  },
  "VERSION" => "0.100",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Tickit/Widget/Progressbar/Vertical.pm
package Tickit::Widget::Progressbar::Vertical;
{
  $Tickit::Widget::Progressbar::Vertical::VERSION = '0.100';
}
use strict;
use warnings;
use parent qw(Tickit::Widget::Progressbar);

=head1 NAME

Tickit::Widget::Progressbar::Vertical - simple progressbar implementation for Tickit

=head1 VERSION

version 0.100

=head1 SYNOPSIS

 my $bar = Tickit::Widget::Progressbar::Vertical->new(
 	completion	=> 0.00,
 );

=head1 DESCRIPTION

See L<Tickit::Widget::Progressbar>.

=cut

use POSIX qw(floor);
use List::Util qw(min);

# Undocumented feature for gradient support. Needs Tickit::Colour,
# since that's not on CPAN then there's little point in enabling this.
use constant ENABLE_GRADIENT => 0;

=head1 METHODS

=cut

sub render_to_rb {
	my $self = shift;
	return $self->render_gradient(@_) if ENABLE_GRADIENT && scalar $self->get_style_values('gradient');
	return $self->render_normal(@_);
}

sub render_gradient {
	my ($self, $rb, $rect) = @_;
	my $win = $self->window or return;

	$rb->clear;
	my $total_height = $win->lines;
	my $cols = $win->cols;
	my $chars = $self->chars;
	my $row = 0;

	my $complete = $self->completion * $total_height;
	my $h = floor($complete);

	my $start = $self->get_style_pen('start');
	my $src = $self->get_style_pen;
	my $dst = Tickit::Pen->new(
		fg => $src->getattr('bg'),
		bg => $src->getattr('fg')
	)->default_from($src);
	my $fg = $src;
	my $bg = $dst;
	while($row < ($total_height - $h)) {
		$rb->goto($row++, 0);
		if($self->direction) {
#			$bg = $self->pen_for_position($row, $total_height, bg => $src, $dst);
			$bg = $self->pen_for_position(
				pos => $total_height - $row,
				total => $total_height,
				from => 'fg',
				to => 'bg',
				start => $start,
				end => $src
			);
			$rb->erase($cols, $bg);
		} else {
#			$fg = $self->pen_for_position($row, $total_height, fg => $src, $dst);
			$rb->erase($cols, $fg);
		}
	}

	if(my $partial = ($complete - $h) * @$chars) {
		if($self->direction) {
			$fg = $self->pen_for_position(
				pos      => $total_height - $row,
				total    => $total_height,
				from     => 'fg',
				to       => 'bg',
				start    => $start,
				end      => $src,
				extra_fg => $dst->getattr('fg'),
			);
		} else {
			$fg = $self->pen_for_position(
				pos => $row,
				total => $total_height,
				from => 'fg',
				start => $start,
				end => $src
			);
		}
#		$bg = $self->pen_for_position($row, $total_height, bg => $src, $dst);
		$rb->char_at($row++, $_, $chars->[$partial], $fg) for 0..$cols - 1;
	}

	while($row <= $total_height) {
		$rb->goto($row, 0);
		if($self->direction) {
#			$fg = $self->pen_for_position($row, $total_height, fg => $src, $dst);
			$rb->erase($cols, $src);
		} else {
			$fg = $self->pen_for_position(
				pos => $row,
				total => $total_height,
				from => 'fg',
				to => 'bg',
				start => $start,
				end => $src
			);
#			$bg = $self->pen_for_position($row, $total_height, fg => $src, $dst);
#			$fg = $self->pen_for_position($row, $total_height, bg => $src, $start);
			$rb->erase($cols, $fg);
		}
		++$row;
	}
}

sub render_normal {
	my ($self, $rb, $rect) = @_;
	my $win = $self->window or return;

	$rb->clear;
	my $total_height = $win->lines;
	my $cols = $win->cols;
	my $chars = $self->chars;
	my $row = 0;

	my $complete = $self->completion * $total_height;
	my $h = floor($complete);

	my $fg = $self->get_style_pen;
	my $bg = Tickit::Pen->new(
		fg => $fg->getattr('bg'),
		bg => $fg->getattr('fg')
	)->default_from($fg);
	while($row < ($total_height - $h)) {
		$rb->goto($row++, 0);
		if($self->direction) {
			$rb->text(' ' x $cols, $bg);
		} else {
			$rb->erase($cols, $fg);
		}
	}

	if(my $partial = ($complete - $h) * @$chars) {
		$rb->char_at($row++, $_, $chars->[$partial], $self->direction ? $bg : $fg) for 0..$cols - 1;
	}

	while($row <= $total_height) {
		$rb->goto($row++, 0);
		if($self->direction) {
			$rb->erase($cols, $fg);
		} else {
			$rb->text(' ' x $cols, $bg);
		}
	}
}

=head2 chars

Returns a list of chars for the various styles we support.

Currently only handles 'ascii' and 'boxchar'.

TODO - this should probably be aligned with the naming
scheme used in other widgets?

=cut

sub chars {
	my $self = shift;
	return {
		ascii	=> [map ord, qw(_ x X)],
		boxchar	=> [
			0x2581,
			0x2582,
			0x2583,
			0x2584,
			0x2585,
			0x2586,
			0x2587,
			0x2588
		],
	}->{$self->style};
}

sub position_for {
	my $self = shift;
	return $self->window->lines - floor(shift() * $self->window->lines);
}

sub expose_between_values {
	my $self = shift;
	return $self unless $self->window;

	my ($prev, $next) = map $self->position_for($_), @_;
	$self->window->expose(
		Tickit::Rect->new(
			top  => min($prev, $next) - 1,
			left => 0,
			cols => $self->window->cols,
			lines => abs($next - $prev) + 1,
		)
	);
}

1;

__END__

=head1 SEE ALSO

=over 4

=item * L<Tickit::Widget::SparkLine>

=back

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2011-2013. Licensed under the same terms as Perl itself.


lib/Tickit/Widget/Progressbar/Horizontal.pm
package Tickit::Widget::Progressbar::Horizontal;
{
  $Tickit::Widget::Progressbar::Horizontal::VERSION = '0.100';
}
use strict;
use warnings;
use parent qw(Tickit::Widget::Progressbar);

=head1 NAME

Tickit::Widget::Progressbar::Horizontal - simple progressbar implementation for Tickit

=head1 VERSION

version 0.100

=head1 SYNOPSIS

 my $bar = Tickit::Widget::Progressbar::Horizontal->new(
 	completion	=> 0.00,
 );

=head1 DESCRIPTION

See L<Tickit::Widget::Progressbar>.

=cut

use POSIX qw(floor);
use List::Util qw(min);

# Undocumented feature for gradient support. Needs Tickit::Colour,
# since that's not on CPAN then there's little point in enabling this.
use constant ENABLE_GRADIENT => 0;

=head1 METHODS

=cut

sub render_to_rb {
	my $self = shift;
	return $self->render_gradient(@_) if ENABLE_GRADIENT && scalar $self->get_style_values('gradient');
	return $self->render_normal(@_);
}

sub render_gradient {
	my ($self, $rb, $rect) = @_;
	my $win = $self->window or return;

	$rb->clear;

	my $total_width = $win->cols;
	my $chars = $self->chars;
	my $complete = $self->completion * $total_width;

	my $start = $self->get_style_pen('start');
	my $src = $self->get_style_pen;
	my $dst = Tickit::Pen->new(
		fg => $src->getattr('bg'),
		bg => $src->getattr('fg')
	)->default_from($src);
	my $fg = $src;
	my $bg = $dst;
	foreach my $line (0..$win->lines - 1) {
		$rb->goto($line, 0);
		my $w = floor($complete);
		if($self->direction) {
			$rb->erase($w, $fg);
		} else {
			for(0..$w-1) {
				$fg = $self->pen_for_position(
					pos => $_,
					total => $total_width,
					from => 'fg',
					to => 'bg',
					start => $src,
					end => $start,
				);
				$rb->erase_at($line, $_, 1, $fg);# for 0..$w-1;
			}
			# $rb->char_at($line, $_, $chars->[-1], $fg) for 0..$w-1;
		}
		if(my $partial = ($complete - $w) * @$chars) {
#			++$w;
			$fg = $self->pen_for_position(
				pos => $w,
				total => $total_width,
				from => 'fg',
				end => $start,
				start => $src
			);
			$rb->char_at($line, $w, $chars->[$partial], $self->direction ? $bg : $fg);
		}
		unless($w >= $total_width) {
			if($self->direction) {
				$rb->char_at($line, $w + $_, $chars->[-1], $fg) for 1..($total_width - $w - 1);
			} else {
				$rb->erase_at($line, $w+1, $total_width - $w, $src);
			}
		}
	}
}

sub render_normal {
	my ($self, $rb, $rect) = @_;
	my $win = $self->window or return;

	$rb->clear;

	my $total_width = $win->cols;
	my $chars = $self->chars;
	my $complete = $self->completion * $total_width;

	# Generate a reversed version of the real pen,
	# since reverse video does not give accurate colours
	# for some terminals (gnome-terminal for example).
	my $fg = $self->get_style_pen;
	my $bg = Tickit::Pen->new(
		fg => $fg->getattr('bg'),
		bg => $fg->getattr('fg')
	)->default_from($fg);

	foreach my $line (0..$win->lines - 1) {
		$rb->goto($line, 0);
		my $w = floor($complete);
		if($self->direction) {
			$rb->erase($w, $fg);
		} else {
			$rb->erase_at($line, 0, $w, $bg);# for 0..$w-1;
			# $rb->char_at($line, $_, $chars->[-1], $fg) for 0..$w-1;
		}
		if(my $partial = ($complete - $w) * @$chars) {
#			++$w;
			$rb->char_at($line, $w, $chars->[$partial], $self->direction ? $bg : $fg);
		}
		unless($w >= $total_width) {
			if($self->direction) {
				$rb->char_at($line, $w + $_, $chars->[-1], $fg) for 1..($total_width - $w - 1);
			} else {
				$rb->erase_at($line, $w+1, $total_width - $w, $fg);
			}
		}
	}
}

=head2 chars

Returns a list of chars for the various styles we support.

Currently only handles 'ascii' and 'boxchar'.

TODO - this should probably be aligned with the naming
scheme used in other widgets?

=cut

sub chars {
	my $self = shift;
	return {
		ascii	=> [qw(| X)],
		boxchar	=> [
			0x258F,
			0x258E,
			0x258D,
			0x258C,
			0x258B,
			0x258A,
			0x2589,
			0x2588,
		],
	}->{$self->style};
}

sub position_for {
	my $self = shift;
	# return $self->window->cols - floor(shift() * $self->window->cols);
	return floor(shift() * $self->window->cols);
}

sub expose_between_values {
	my $self = shift;
	return $self unless $self->window;

	my ($prev, $next) = map $self->position_for($_), @_;
	$self->window->expose(
		Tickit::Rect->new(
			left  => min($prev, $next) - 1,
			top => 0,
			lines => $self->window->lines,
			cols => abs($next - $prev) + 2,
		)
	);
}

1;

__END__

=head1 SEE ALSO

=over 4

=item * L<Tickit::Widget::SparkLine>

=back

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2011-2013. Licensed under the same terms as Perl itself.


lib/Tickit/Widget/Progressbar.pm
package Tickit::Widget::Progressbar;
# ABSTRACT: horizontal/vertical progress bars for Tickit
use strict;
use warnings;
use parent qw(Tickit::Widget);

our $VERSION = '0.100';

=head1 NAME

Tickit::Widget::Progressbar - simple progressbar implementation for Tickit

=head1 VERSION

version 0.100

=head1 SYNOPSIS

 use Tickit::Widget::Progressbar::Horizontal;
 my $bar = Tickit::Widget::Progressbar::Horizontal->new(
 	completion	=> 0.00,
 );
 $bar->completion($_ / 100.0) for 0..100;

=head1 DESCRIPTION

Provides support for a 'progress bar' widget. Use the L<Tickit::Widget::Progressbar::Horizontal>
or L<Tickit::Widget::Progressbar::Vertical> subclasses depending on whether you want the progress
bar to go from left to right or bottom to top.

=cut

use Tickit::Style;

use constant CLEAR_BEFORE_RENDER => 0;
use constant WIDGET_PEN_FROM_STYLE => 0;
use constant CAN_FOCUS => 0;

BEGIN {
	style_definition base =>
		fg => 255,
		bg => 'black',
		gradient => 0,
		start_fg => 232;
}

=head1 METHODS

=cut

sub lines { 1 }
sub cols { 1 }

=head2 new

Instantiate a new L<Tickit::Widget::Progressbar> object. Takes the following named parameters:

=over 4

=item * completion - a value from 0.0 to 1.0 indicating progress

=item * orientation - 'vertical' or 'horizontal'

=item * direction - whether progress goes forwards (left to right, bottom to top) or backwards
(right to left, top to bottom).

=back

Note that this is a base class, and the appropriate L<Tickit::Widget::Progressbar::Horizontal>
or L<Tickit::Widget::Progressbar::Vertical> subclass should be used when instantiating a real
widget.

=cut

sub new {
	my $class = shift;
	my %args = @_;
	my $completion = delete $args{completion};
	my $orientation = delete $args{orientation};
	my $direction = delete $args{direction};
	my $self = $class->SUPER::new(%args);
	$self->{completion} = $completion || 0.0;
	$self->{orientation} = $orientation || 'horizontal';
	$self->{direction} = $direction || 0;
	return $self;
}

sub orientation { 'horizontal' }
sub style { 'boxchar' }
sub direction { shift->{direction} }

=head2 completion

Accessor for the current progress bar completion state - call this with a float value from 0.00..1.00
to set completion and re-render.

=cut

sub completion {
	my $self = shift;
	if(@_) {
		my $previous = $self->{completion};
		$self->{completion} = shift;
		if(defined $previous) {
			# Not entirely sure this part is working reliably enough yet
#			$self->expose_between_values($previous, $self->{completion});
			$self->redraw;
		} else {
			$self->redraw;
		}
		return $self;
	}
	return $self->{completion};
}

sub pen_for_position {
	my $self = shift;
	my %args = @_;
	$self->{gradient_pen} ||= {};
	$self->{gradient_pen}{join ',', map { $_ => $args{$_} } sort keys %args} ||= do {
		my @start = Tickit::Colour->colour_to_rgb($args{start}->getattr($args{from}));
		my @end = Tickit::Colour->colour_to_rgb($args{end}->getattr($args{from}));
		my $col = Tickit::Colour->rgb_to_colour(map {
				($start[$_] * $args{pos} + $end[$_] * ($args{total} - $args{pos})) / $args{total}
			} 0..2);
		my %extra = map {; /^extra_(.*)$/ ? ($1 => $args{$_}) : () } keys %args;

		Tickit::Pen::Immutable->new(
			($args{to} || $args{from}) => $col,
			%extra,
		);
	};
}

1;

__END__

=head1 SEE ALSO

L<Tickit>

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2011-2013. Licensed under the same terms as Perl itself.
